app/services/order_cycle_distributed_products.rb

```diff
@@ -13,6 +13,13 @@ def products_relation
     Spree::Product.where(id: stocked_products).group("spree_products.id")
   end
 
+  def products_taxons_relation
+    Spree::Product.where(id: stocked_products).
+      joins("LEFT JOIN (SELECT DISTINCT ON(product_id) id, product_id, primary_taxon_id FROM spree_variants) first_variant ON spree_products.id = first_variant.product_id").
```

Comment:
I think we can try using left_joins to make this slightly shorter:
```suggestion
      left_joins("(SELECT DISTINCT ON(product_id) id, product_id, primary_taxon_id FROM spree_variants) first_variant ON spree_products.id = first_variant.product_id").
```
We'll also need to fix the linter error.

I haven't got any better ideas right now sorry..

Reply:
I couldn't get `left_joins` to work, so have kept it as-is (with added line-breaks for code formatting).